src/core/etl/tests/Flow/ETL/Tests/Unit/Extractor/FlowTestCase.php

```diff
@@ -0,0 +1,71 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Flow\ETL\Tests\Unit\Extractor;
+
+use Flow\ETL\{Config, Extractor, FlowContext, Rows};
+use PHPUnit\Framework\TestCase;
+
+abstract class FlowTestCase extends TestCase
+{
+    public function assertExtractorCountBatches(int $expectedCount, Extractor $extractor) : void
+    {
+        static::assertCount(
+            $expectedCount,
+            \iterator_to_array($extractor->extract(new FlowContext(Config::default())))
+        );
+    }
+
+    public function assertExtractorCountRows(int $expectedCount, Extractor $extractor) : void
+    {
+        $totalRows = 0;
+
+        foreach ($extractor->extract(new FlowContext(Config::default())) as $rows) {
+            $totalRows += $rows->count();
+        }
+
+        static::assertSame($expectedCount, $totalRows);
+    }
+
+    public function assertExtractorCountRowsPerBatch(int $expectedCount, Extractor $extractor) : void
+    {
+        $extractorContainsBatches = false;
+
+        foreach ($extractor->extract(new FlowContext(Config::default())) as $rows) {
+            static::assertCount($expectedCount, $rows);
+            $extractorContainsBatches = true;
+        }
+
+        if (!$extractorContainsBatches) {
+            static::fail('Extractor does not contain any batches');
+        }
+    }
+
+    /**
+     * @param array<Rows> $expectedRows
+     * @param Extractor $extractor
+     */
+    public function assertExtractorEqualsRows(array $expectedRows, Extractor $extractor) : void
```

Comment:
maybe `assertExtractorRowsEquals(Rows $rows, Extractor $extractor, ?string $message = null) : void`

and under the hood what you can do is merge all the rows extracted by the extractor:

```php
$extractedRows = rows();

foreach ($extractor->extract(flow_context()) as $nextRows) {
    $extractedRows = $extractedRows->merge($nextRows);
}

self::assertEquals($expextedRows, $extractedRows);
```

Reply:
and I would then make a similar method `assertExtractRowsAsArrayEquals`

Reply:
Fixed in https://github.com/flow-php/flow/pull/1291/commits/315d9f96f0dbb6bd19031ea43392d183f53a4173 and https://github.com/flow-php/flow/pull/1291/commits/edef63a70f5c86bbeaa1a197a3439077ffc83390